Userland/Libraries/LibCore/BufferingIODevice.h

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (c) 2021, Ali Mohammad Pur <mpfard@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/TypedTransfer.h>
+#include <LibCore/FileLikeIODevice.h>
+
+namespace Core {
+// This is not necessarily a valid iterator in all contexts,
+// if we had concepts, this would be InputIterator, not Copyable, Movable.
+template<typename UnderlyingDeviceT>
+class LineIterator {
+    AK_MAKE_NONCOPYABLE(LineIterator);
+
+public:
+    explicit LineIterator(BufferingIODevice<UnderlyingDeviceT>& device, bool is_end = false)
+        : m_device(device)
+        , m_is_end(is_end)
+    {
+        ++*this;
+    }
+
+    bool operator==(const LineIterator& other) const { return &other == this || (at_end() && other.is_end()) || (other.at_end() && is_end()); }
+    bool is_end() const { return m_is_end; }
+    bool at_end() const;
+
+    LineIterator& operator++();
+
+    StringView operator*() const { return m_buffer; }
+
+private:
+    NonnullRefPtr<BufferingIODevice<UnderlyingDeviceT>> m_device;
+    bool m_is_end { false };
+    String m_buffer;
+};
+
+template<typename UnderlyingDeviceT>
+class BufferingIODevice : public UnderlyingDeviceT {
+    C_OBJECT(BufferingIODevice)
```

Comment:
Let's put an `;` at the end of the line for consistency with the rest of the codebase.